Documentation

danielaRiesgo opened the pull request:
## Sumary ##
**Trello card**: https://trello.com/c/gFXtp9T0/32-documentar

Documenting all framework elements.
And some small refactoring that came up:
    Changed DefaultXXXViewConfiguration, to a XXXViewConfiguration struct with default init parameters.
    Added/Removed view optional elements that were missing/unnecessary.

## Test Plan ##

  * Test all the app to check nothing changed.
  * Both login and signup succeed if using (email: "example@mail.com", password: "password")

## Known issues ##
Some session service issues are left as TODO because decisions are to be made (in another PR) about it.

woloxci commented:
:+1: There are no linter violations.

guidomb commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -63,10 +64,10 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
 
-    /**
+    /*
         Bootstraps your project with the authentication framework,
```
What about something like this?

```
Bootstraps your application with the authentication framework,
starting with the login screen if no user is logged in.
Otherwise, it runs your application's Main View Controller.
```

danielaRiesgo replied:
The only thing  is that then it will start with login or signup according to what the user decides. `authentication process`? Or maybe explaining exactly that: `starting with the login or signup screen (according to you preferences) if no user is logged in.`

guidomb commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -201,52 +294,95 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
+         It will pop the signup controller from the navigation controller,
+         to return to login screen.
+     */
     public func onLogin(controller: SignupController) {
```
this method should be final

guidomb commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -201,52 +294,95 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
+         in the login screen.
+         It will push the new controller in the navigation controller.
+     */
     public func onRecoverPassword(controller: LoginController) {
```
this method should be final

guidomb commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -201,52 +294,95 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
+         login screen.
+         It will push the new controller in the navigation controller.
+     */
     public func onSignup(controller: LoginController) {
```
this method should be final

guidomb commented on Authentication/Login/Model/User.swift:
```diff
@@ -8,7 +8,11 @@
+     a user for it to be compatible with
+     the authentication framework.
+ */
 public protocol UserType {
```
Maybe for another PR but I think we can remove this protocol.

guidomb commented on Authentication/Login/View/LoginController.swift:
```diff
@@ -12,13 +12,19 @@ import Rex
+    presents only the first one in the errors label.
+ 
+    If wanting to use the default LoginController with some customization,
+    you should not override the `createLoginController` method of the Bootstrapper,
```
Users cannot override `createLoginController` because its visibility is `internal`.

guidomb replied:
Also I don't think this should be explained here because all those methods are part of the bootstrapper. I think that the documentation should only "talk" about the class being documented not how other classes use it.

danielaRiesgo replied:
I agree. But as a framework I think this kind of explanations are nice to have, although maybe this is not the place. Where could we move it?

guidomb replied:
Yeah that is true. My main concern is not being "coupled" with other parts of the library in a method or class documentation because, most probably, if something on how this class is used changes (not necessarily this class' API) then this docs would get out of date. Maybe the best place for high level usage documentation is the README, a tutorial or the Wiki.

guidomb commented on Authentication/Login/View/LoginViewConfiguration.swift:
```diff
@@ -6,8 +6,12 @@
-import Foundation
-
+/*
+    Represents the configurations able to
```
What about `Represents the configurable parameters of a view that conforms to LoginViewType`?

guidomb commented:
I've made a quick pass and left some comments. Haven't got the time to do deep review. I guess that the documentation will improve over time with the usage of the library so I think it is a very good starting point. We should see if other devs can properly configure the lib without much guidance and only using the docs to be sure that they are good.

woloxci commented:
:+1: There are no linter violations.

woloxci commented:
Total linter violations in pull request: 1
Serious: 0
Warnings: 1

woloxci commented on Authentication/SignUp/View/SignupController.swift:
```diff
@@ -8,6 +8,21 @@
+     If there are more than one validation error in a field, the controller
+     presents only the first one in the errors label.
+     
+     If wanting to use the default SignupController, you should not override
```
WARNING - TODOs and FIXMEs should be avoided.

FranDepascuali commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -8,7 +8,7 @@
 import Foundation
 
-/**
+/*
```
Why did you change this? If you remove the second *, the user of this class will not be able to see the description.

FranDepascuali replied:
with `/**`:
<img width="261" alt="screen shot 2016-07-27 at 21 51 24" src="https://cloud.githubusercontent.com/assets/12101394/17197591/54e01d78-5444-11e6-9e8e-4094dc18428c.png">

FranDepascuali replied:
with `/*`( just one *):
<img width="261" alt="screen shot 2016-07-27 at 21 52 36" src="https://cloud.githubusercontent.com/assets/12101394/17197607/75214012-5444-11e6-996c-8e96c1410f81.png">

FranDepascuali commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -31,7 +34,7 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
     }
 
-    /**
+    /*
```
Why did you change this? If you remove the second *, the user of this class will not be able to see the description.

FranDepascuali commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -75,49 +76,87 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
+}
 
-    /**
+// MARK: - Login Functions
```
You could leave mark, but extensions already make it clear

FranDepascuali commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -75,49 +76,87 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
+// MARK: - Login Functions
+public extension AuthenticationBootstrapper {
+    
+    /*
```
Why did you change this? If you remove the second *, the user of this class/function will not be able to see the description.

FranDepascuali commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -75,49 +76,87 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
+        return LoginController(configuration: configuration)
+    }
+    
+    /*
```
Why did you change this? If you remove the second *, the user of this class/function will not be able to see the description.

FranDepascuali commented on Authentication/AuthenticationBootstrapper.swift:
```diff
@@ -75,49 +76,87 @@ public class AuthenticationBootstrapper<User: UserType, SessionService: SessionS
 
-    /**
-         Creates the LogInViewModel to use in the authentication process logic,
+    /*
```
Why did you change this? If you remove the second *, the user of this class/function will not be able to see the description.

FranDepascuali commented:
@danielaRiesgo there are a lot of places in which you changed /** to /*. This are different things: with two **, it is documented (meaning that everyone who uses that function/class will be able to see the description). With just one *, it is just a comment, so you will only be able to see it if you are reading the implementation. I think you want the two ** (`/**`)

FranDepascuali commented on Authentication/ViewProtocols.swift:
```diff
@@ -6,6 +6,10 @@
+/*
+     Represents something that can give us an
+     UIView to show and render.
+ */
```
I would write just: 
```swift
Represents a component that can be visually displayed on the screen
```

FranDepascuali commented:
@danielaRiesgo 
The only small detail is that there are some places in which you describe, for example a protocol, with something that it exposes or some detail about its implementation.
I think that those particular details should go on the things that it exposes and not on the protocol itself. The comment for a protocol should just describe what it does (or means). 

Having that said, it is just a small detail (and I don't even know if I'm right), but I like very much how you document functions, protocols and classes at all. I think that you are one of the best of us in that sense, very good 👍

danielaRiesgo commented:
I changed to `/*` because I saw it maid valid_docs violations disappear. Now I understand why. I changed it back, and tried to solve all violations, but there are two I can't figure out what the problem is.

danielaRiesgo commented:
@FranDepascuali there I checked the protocols' documentation, and tried to improve it, but I don't know if I got what you meant.

woloxci commented:
Total linter violations in pull request: 1
Serious: 0
Warnings: 1

FranDepascuali commented on AuthenticationDemo/en.lproj/Main.strings:
```diff
@@ -1,3 +1,3 @@
 
-/* Class = "UILabel"; text = "Esta es la pantalla de tu aplicación !"; ObjectID = "VF1-gk-0yi"; */
+/** Class = "UILabel"; text = "Esta es la pantalla de tu aplicación !"; ObjectID = "VF1-gk-0yi"; */
```
I think you want this to be a comment

FranDepascuali commented:
correct that detail and 👍

woloxci commented:
Total linter violations in pull request: 1
Serious: 0
Warnings: 1

guidomb commented:
@danielaRiesgo Did you rebase this PR? If so try not to do that once a review is in progress. If think I've already approved this one so if you just fixed the comments made by @FranDepascuali is ready to be merged. I am not reviewing this again so I trust that is the case. 👍
